File: android/app/src/main/kotlin/com/brandbuilder/openpos/StarPrinterChannel.kt
package com.brandbuilder.openpos

import android.content.Context
import android.graphics.BitmapFactory
import android.os.Handler
import android.os.Looper
import com.starmicronics.stario10.InterfaceType
import com.starmicronics.stario10.StarConnectionSettings
import com.starmicronics.stario10.StarDeviceDiscoveryManager
import com.starmicronics.stario10.StarDeviceDiscoveryManagerFactory
import com.starmicronics.stario10.StarPrinter
import com.starmicronics.stario10.starxpandcommand.DocumentBuilder
import com.starmicronics.stario10.starxpandcommand.PrinterBuilder
import com.starmicronics.stario10.starxpandcommand.StarXpandCommandBuilder
import com.starmicronics.stario10.starxpandcommand.printer.CutType
import com.starmicronics.stario10.starxpandcommand.printer.ImageParameter
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Bridge to Star's StarXpand SDK for TSP100III printers over classic
 * Bluetooth. Deliberately thin: discovery and printing an already-rendered
 * image, nothing else.
 *
 * The TSP100III is graphics-only — it cannot be sent text — so the Dart side
 * rasterises the receipt and passes PNG pages down. See
 * `lib/features/printing/data/star_printer.dart`.
 *
 * Only the Bluetooth interface is enabled here. The app is offline by
 * design, so LAN/Wi-Fi printer discovery is intentionally never requested.
 */
class StarPrinterChannel(
    private val context: Context,
    messenger: BinaryMessenger,
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private val methods = MethodChannel(messenger, METHOD_CHANNEL).apply {
        setMethodCallHandler(this@StarPrinterChannel)
    }
    private val discoveryEvents = EventChannel(messenger, EVENT_CHANNEL).apply {
        setStreamHandler(this@StarPrinterChannel)
    }

    private val main = Handler(Looper.getMainLooper())
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var manager: StarDeviceDiscoveryManager? = null
    private var sink: EventChannel.EventSink? = null

    fun dispose() {
        stopDiscovery()
        methods.setMethodCallHandler(null)
        discoveryEvents.setStreamHandler(null)
        scope.cancel()
    }

    // ── EventChannel.StreamHandler ────────────────────────────────────────

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    // ── MethodChannel.MethodCallHandler ───────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startDiscovery" -> startDiscovery(result)
            "stopDiscovery" -> {
                stopDiscovery()
                result.success(null)
            }
            "printImages" -> printImages(call, result)
            else -> result.notImplemented()
        }
    }

    private fun startDiscovery(result: MethodChannel.Result) {
        stopDiscovery()
        try {
            val found = StarDeviceDiscoveryManagerFactory.create(
                listOf(InterfaceType.Bluetooth),
                context,
            )
            found.discoveryTime = DISCOVERY_MILLIS
            found.callback = object : StarDeviceDiscoveryManager.Callback {
                override fun onPrinterFound(printer: StarPrinter) {
                    val model = printer.information?.model?.name
                    main.post {
                        sink?.success(
                            mapOf(
                                "address" to printer.connectionSettings.identifier,
                                "name" to (model ?: "Star printer"),
                            ),
                        )
                    }
                }

                override fun onDiscoveryFinished() {
                    main.post { sink?.success(mapOf("finished" to true)) }
                }
            }
            found.startDiscovery()
            manager = found
            result.success(null)
        } catch (e: Throwable) {
            // Bluetooth off, permission not granted, SDK failure — the Dart
            // side turns this into a snackbar.
            result.error("discovery_failed", e.message ?: e.toString(), null)
        }
    }

    private fun stopDiscovery() {
        manager?.stopDiscovery()
        manager = null
    }

    private fun printImages(call: MethodCall, result: MethodChannel.Result) {
        val address = call.argument<String>("address")
        val images = call.argument<List<ByteArray>>("images")
        val width = call.argument<Int>("width") ?: DEFAULT_WIDTH_DOTS
        if (address.isNullOrEmpty() || images.isNullOrEmpty()) {
            result.error("bad_args", "address and images are required", null)
            return
        }

        scope.launch {
            val printer = StarPrinter(
                StarConnectionSettings(InterfaceType.Bluetooth, address),
                context,
            )
            try {
                val page = PrinterBuilder()
                for (bytes in images) {
                    val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                        ?: throw IllegalArgumentException("Receipt image could not be decoded")
                    page.actionPrintImage(ImageParameter(bitmap, width))
                }
                page.actionCut(CutType.Partial)

                val commands = StarXpandCommandBuilder()
                    .addDocument(DocumentBuilder().addPrinter(page))
                    .getCommands()

                printer.openAsync().await()
                printer.printAsync(commands).await()
                main.post { result.success(null) }
            } catch (e: Throwable) {
                main.post {
                    result.error("print_failed", e.message ?: e.toString(), null)
                }
            } finally {
                runCatching { printer.closeAsync().await() }
            }
        }
    }

    private companion object {
        const val METHOD_CHANNEL = "pos_app/star_printer"
        const val EVENT_CHANNEL = "pos_app/star_printer/devices"
        const val DISCOVERY_MILLIS = 10_000
        /** 72mm of printable width at 203 dpi — an 80mm roll. */
        const val DEFAULT_WIDTH_DOTS = 576
    }
}

File: ios/Runner/StarPrinterChannel.swift
import Flutter
import StarIO10
import UIKit

/// Bridge to Star's StarXpand SDK for TSP100III printers over Bluetooth.
/// Mirrors `android/app/src/main/kotlin/.../StarPrinterChannel.kt`:
/// discovery and printing an already-rendered image, nothing else.
///
/// The TSP100III is graphics-only — it cannot be sent text — so the Dart
/// side rasterises the receipt and passes PNG pages down. See
/// `lib/features/printing/data/star_printer.dart`.
///
/// On iOS the printer is an MFi accessory: it must be paired in Settings ›
/// Bluetooth, and the app must declare `jp.star-m.starpro` under
/// `UISupportedExternalAccessoryProtocols` in Info.plist, or discovery
/// finds nothing.
///
/// Only the Bluetooth interface is enabled here. The app is offline by
/// design, so LAN/Wi-Fi printer discovery is intentionally never requested.
final class StarPrinterChannel: NSObject {
    private static let methodChannelName = "pos_app/star_printer"
    private static let eventChannelName = "pos_app/star_printer/devices"
    private static let discoveryMillis = 10_000
    /// 72mm of printable width at 203 dpi — an 80mm roll.
    private static let defaultWidthDots = 576

    private let methods: FlutterMethodChannel
    private let discoveryEvents: FlutterEventChannel
    private var manager: (any StarDeviceDiscoveryManager)?
    private var sink: FlutterEventSink?

    init(messenger: FlutterBinaryMessenger) {
        methods = FlutterMethodChannel(
            name: Self.methodChannelName, binaryMessenger: messenger)
        discoveryEvents = FlutterEventChannel(
            name: Self.eventChannelName, binaryMessenger: messenger)
        super.init()
        methods.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result: result)
        }
        discoveryEvents.setStreamHandler(self)
    }

    func dispose() {
        stopDiscovery()
        methods.setMethodCallHandler(nil)
        discoveryEvents.setStreamHandler(nil)
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startDiscovery":
            startDiscovery(result)
        case "stopDiscovery":
            stopDiscovery()
            result(nil)
        case "printImages":
            printImages(call.arguments as? [String: Any] ?? [:], result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // ── Discovery ─────────────────────────────────────────────────────────

    private func startDiscovery(_ result: @escaping FlutterResult) {
        stopDiscovery()
        do {
            let found = try StarDeviceDiscoveryManagerFactory.create(
                interfaceTypes: [InterfaceType.bluetooth])
            found.discoveryTime = Self.discoveryMillis
            found.delegate = self
            try found.startDiscovery()
            manager = found
            result(nil)
        } catch {
            // Bluetooth off, accessory protocol not declared, SDK failure —
            // the Dart side turns this into a snackbar.
            result(FlutterError(
                code: "discovery_failed",
                message: "\(error)",
                details: nil))
        }
    }

    private func stopDiscovery() {
        manager?.stopDiscovery()
        manager = nil
    }

    // ── Printing ──────────────────────────────────────────────────────────

    private func printImages(_ args: [String: Any], _ result: @escaping FlutterResult) {
        guard let address = args["address"] as? String, !address.isEmpty,
              let pages = args["images"] as? [FlutterStandardTypedData], !pages.isEmpty
        else {
            result(FlutterError(
                code: "bad_args",
                message: "address and images are required",
                details: nil))
            return
        }
        let width = args["width"] as? Int ?? Self.defaultWidthDots

        let images = pages.compactMap { UIImage(data: $0.data) }
        guard images.count == pages.count else {
            result(FlutterError(
                code: "print_failed",
                message: "Receipt image could not be decoded",
                details: nil))
            return
        }

        let page = StarXpandCommand.PrinterBuilder()
        for image in images {
            _ = page.actionPrintImage(
                StarXpandCommand.Printer.ImageParameter(image: image, width: width))
        }
        _ = page.actionCut(StarXpandCommand.Printer.CutType.partial)

        let commands = StarXpandCommand.StarXpandCommandBuilder()
            .addDocument(StarXpandCommand.DocumentBuilder().addPrinter(page))
            .getCommands()

        let printer = StarPrinter(
            StarConnectionSettings(interfaceType: .bluetooth, identifier: address))

        Task {
            do {
                try await printer.open()
                try await printer.print(command: commands)
                await printer.close()
                Self.onMain { result(nil) }
            } catch {
                await printer.close()
                Self.onMain {
                    result(FlutterError(
                        code: "print_failed",
                        message: "\(error)",
                        details: nil))
                }
            }
        }
    }

    /// Channel replies and events must reach Flutter on the platform thread.
    private static func onMain(_ work: @escaping () -> Void) {
        if Thread.isMainThread {
            work()
        } else {
            DispatchQueue.main.async(execute: work)
        }
    }
}

// ── Discovery results ─────────────────────────────────────────────────────

extension StarPrinterChannel: StarDeviceDiscoveryManagerDelegate {
    func manager(_ manager: any StarDeviceDiscoveryManager, didFind printer: StarPrinter) {
        let identifier = printer.connectionSettings.identifier
        let model = printer.information.map { String(describing: $0.model) }
        Self.onMain { [weak self] in
            self?.sink?([
                "address": identifier,
                "name": model ?? "Star printer",
            ])
        }
    }

    func managerDidFinishDiscovery(_ manager: any StarDeviceDiscoveryManager) {
        Self.onMain { [weak self] in
            self?.sink?(["finished": true])
        }
    }
}

// ── Discovery event stream ────────────────────────────────────────────────

extension StarPrinterChannel: FlutterStreamHandler {
    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        sink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
